Adds semanticsLabel to MenuItemButton

philipfranchi opened the pull request:
This PR adds a `semanticsLabel` field to the MenuItemButton, letting users pass in a Semantics Label if they have issues with the way a ScreenReader reads the generated text.

fixes: [142939: Screen Readers won't read punctuation characters in Shortcuts](https://github.com/flutter/flutter/issues/142939)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

philipfranchi committed:
Adds semanticsLabel to MenuItemButton

philipfranchi committed:
Merge branch 'master' into add_semantics_label_to_menu_item_button

philipfranchi committed:
Added const keyword to test constructer

philipfranchi committed:
Merge branch 'master' into add_semantics_label_to_menu_item_button

philipfranchi committed:
Added an extra doc comment

philipfranchi committed:
Merge branch 'master' into add_semantics_label_to_menu_item_button

gspencergoog commented on packages/flutter/test/material/menu_anchor_test.dart:
```diff
@@ -3329,6 +3329,27 @@ void main() {
+
+      expect(find.bySemanticsLabel('TestWidget'), findsOneWidget);
+      semantics.dispose();
+    },variant: TargetPlatformVariant.desktop());
```
```suggestion
    }, variant: TargetPlatformVariant.desktop());
```

philipfranchi committed:
Update packages/flutter/test/material/menu_anchor_test.dart

Co-authored-by: Greg Spencer <gspencergoog@users.noreply.github.com>

philipfranchi committed:
fix missing space

engine-flutter-autoroll committed:
Roll Flutter Engine from 9df2d3a0778e to 7c9d5adb6ff8 (3 revisions) (#145909)

https://github.com/flutter/engine/compare/9df2d3a0778e...7c9d5adb6ff8

2024-03-28 skia-flutter-autoroll@skia.org Roll Skia from abd1774b901f to 9e3fe3b324b7 (5 revisions) (flutter/engine#51745)
2024-03-28 skia-flutter-autoroll@skia.org Roll Skia from cef0d40c5a06 to abd1774b901f (1 revision) (flutter/engine#51744)
2024-03-28 skia-flutter-autoroll@skia.org Roll Skia from 66241be7c81c to cef0d40c5a06 (1 revision) (flutter/engine#51743)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC jacksongardner@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://issues.skia.org/issues/new?component=1389291&template=1850622

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

mdebbar committed:
Add `viewId` to `TextInputConfiguration` (#145708)

In order for text fields to work correctly in multi-view on the web, we need to have the `viewId` information sent to the engine (`TextInput.setClient`). And while the text field is active, if it somehow moves to a new `View`, we need to inform the engine about such change (`TextInput.updateConfig`).

Engine PR: https://github.com/flutter/engine/pull/51099
Fixes https://github.com/flutter/flutter/issues/137344

engine-flutter-autoroll committed:
Roll Packages from e6b3e11d9d5a to 924c7e6e899d (5 revisions) (#145915)

https://github.com/flutter/packages/compare/e6b3e11d9d5a...924c7e6e899d

2024-03-28 98614782+auto-submit[bot]@users.noreply.github.com Reverts "[ci] Temporarily allow-warnings in podspec_check_command.dart (#6416)" (flutter/packages#6419)
2024-03-28 ditman@gmail.com [ci] Temporarily allow-warnings in podspec_check_command.dart (flutter/packages#6416)
2024-03-28 98614782+auto-submit[bot]@users.noreply.github.com Reverts "[interactive_media_ads] Creates and adds the `interactive_media_ads` plugin (#6060)" (flutter/packages#6417)
2024-03-27 10687576+bparrishMines@users.noreply.github.com [interactive_media_ads] Creates and adds the `interactive_media_ads` plugin (flutter/packages#6060)
2024-03-27 43054281+camsim99@users.noreply.github.com [camerax] Use `AspectRatioStrategy` to help automatic selection of expected resolution (flutter/packages#6357)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-packages-flutter-autoroll
Please CC flutter-ecosystem@google.com,rmistry@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://issues.skia.org/issues/new?component=1389291&template=1850622

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

TahaTesser committed:
Update `TabBar` and `TabBar.secondary` to use indicator height/color M3 tokens (#145753)

fixes [Secondary `TabBar` indicator height token is missing ](https://github.com/flutter/flutter/issues/124965)

### Code sample

<details>
<summary>expand to view the code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: DefaultTabController(
        length: 2,
        child: Scaffold(
          appBar: AppBar(
            title: const Text('Sample'),
            bottom: const TabBar.secondary(
              tabs: <Widget>[
                Tab(icon: Icon(Icons.directions_car)),
                Tab(icon: Icon(Icons.directions_transit)),
              ],
            ),
          ),
          floatingActionButton: FloatingActionButton(
            onPressed: () {},
            child: const Icon(Icons.add),
          ),
        ),
      ),
    );
  }
}
```

</details>

engine-flutter-autoroll committed:
Roll Flutter Engine from 7c9d5adb6ff8 to 043af350ae85 (2 revisions) (#145917)

https://github.com/flutter/engine/compare/7c9d5adb6ff8...043af350ae85

2024-03-28 skia-flutter-autoroll@skia.org Roll Skia from 4e2205aff93b to c0e0b76d6d51 (2 revisions) (flutter/engine#51750)
2024-03-28 skia-flutter-autoroll@skia.org Roll Skia from 9e3fe3b324b7 to 4e2205aff93b (5 revisions) (flutter/engine#51747)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC jacksongardner@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://issues.skia.org/issues/new?component=1389291&template=1850622

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

sealesj committed:
Refactor skp_generator_tests (#145871)

Refactor skp_generator_tests suite in order to reduce testing logic in test.dart and allow for later implementing package:test onto the existing skp_generator_tests tests

Part of https://github.com/flutter/flutter/issues/145482

engine-flutter-autoroll committed:
Roll Flutter Engine from 043af350ae85 to f71e5ad8586b (1 revision) (#145919)

https://github.com/flutter/engine/compare/043af350ae85...f71e5ad8586b

2024-03-28 dkwingsmt@users.noreply.github.com Make EmbedderExternalViewEmbedder present to the correct view (flutter/engine#51719)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC jacksongardner@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://issues.skia.org/issues/new?component=1389291&template=1850622

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

philipfranchi committed:
Merge branch 'master' into add_semantics_label_to_menu_item_button

philipfranchi committed:
Merge branch 'master' into add_semantics_label_to_menu_item_button

gspencergoog committed:
Merge branch 'master' into add_semantics_label_to_menu_item_button

philipfranchi committed:
Merge branch 'master' into add_semantics_label_to_menu_item_button

philipfranchi committed:
Merge branch 'master' into add_semantics_label_to_menu_item_button

philipfranchi committed:
Merge branch 'master' into add_semantics_label_to_menu_item_button

gspencergoog commented on packages/flutter/lib/src/material/menu_anchor.dart:
```diff
@@ -3029,6 +3044,10 @@ class _MenuItemLabel extends StatelessWidget {
       ],
     );
+    if (semanticsLabel != null) {
+      menuItemLabel = Semantics(label: semanticsLabel, excludeSemantics: true, child: menuItemLabel);
```
Do we want to exclude semantics for all of the children, or just the `widget.child`?  I can see someone wanting to put some kind of button in as part of the trailing or leading icon, and those would want to keep their semantics. (See the "Zoom" line in the Chrome "..." menu for an example).

We could also just assume that people who want to do something that complex will be making their own widget to add to the menu, but they'll probably start by trying to use this, so if it's possible to let them, then we might as well not prevent it.

gspencergoog replied:
Or I suppose they could still do it, they'd just have to not supply a semantics label (and know not to do that).

philipfranchi replied:
That's interesting, I can definitely see that scenario playing out. In our case,  we need to overwrite the semantic label on both the child and the serialized shortcut decorations, while still displaying the generated shortcut icons. 

What if instead of passing in the label, we exposed a field called `optionalSemanticsProperties` and said explicitly in the documentation `Setting these properties overwrites the Semantics of the entire Widget`, or something similar? That seems to me to be a clearer message of "this applies to all descendants" than just a label would.

gspencergoog replied:
Yes, that might work.  But what would that pass in?  Not just a string, right?

 Maybe a better idea would be to have an optional `semanticsBuilder` argument that is a callback that takes a `Widget` and returns a `Widget`. The input to the builder is the child for the semantics widget (`menuItemLabel` in your change), and the output is a `Semantics` widget wrapping the given child. It shouldn't require a `Semantics` widget return value, since there are cases where you might want to wrap it in something else (e.g. an `ExcludeSemantics` widget).

Then you have full control over what the semantics looks like for the widget.

However, at that point, why can't you just wrap the `MenuItemButton` with a `Semantics` widget and get what you want? I seem to vaguely remember discussing this, but I can't remember why it didn't work.

franciscojma86 replied:
I also might not remember exactly, but I think the problem was mostly with SubmenuButton, since wrapping it in a Semantics label resulted in merging the submenu semantics, or replacing them if I'm not mistaken. At a MenuItemButton level, I'm not sure if that doesn't work. I'd need to test that again

gspencergoog replied:
Ahh, yes, that sounds familiar.  I would think that the problem has to do with only the `SubmenuButton`, since that would definitely cause the children's semantics to be lumped together.  You should be able to apply a `Semantics` widget around each child to override their semantics, but I think we even talked about that and decided that was tedious, and we arrived here.  My callback suggestion would definitely get us back to tedious again.

> What if instead of passing in the label, we exposed a field called `optionalSemanticsProperties` and said explicitly in the documentation `Setting these properties overwrites the Semantics of the entire Widget`, or something similar? That seems to me to be a clearer message of "this applies to all descendants" than just a label would.

How about if we go back to this, except I'd call it `semanticsLabel` as before, since making it nullable should imply optional anyhow, and it's just the label, not the other properties. Make clear that it's just the label in the docs too, and maybe mention that if they want to override other semantics properties, they can wrap the item in a `Semantics` widget.

philipfranchi replied:
Cool, so I can update the docs here to let the user know they can wrap this in Semantics. But I'm a little confused about the difference between this proposal and the existing PR? It seems like that's what this code is doing. The main addition is the extra `excludeSemantics` field which is there to prevent the Screen Reader from reading beyond what was given in the label. Beyond that, which part of the PR should I tweak?

gspencergoog replied:
Sorry, I wasn't too clear there. I guess this PR is doing what I was I was asking for, except for the docs.

philipfranchi committed:
Documents how setting label overwrites semantics.

gspencergoog commented on packages/flutter/lib/src/material/menu_anchor.dart:
```diff
@@ -894,8 +894,13 @@ class MenuItemButton extends StatefulWidget {
   ///
-  /// A screen reader will default to reading the genereated text on the [MenuItemButton]
-  /// itself, which is not guaranteed to be readable.
+  /// A screen reader will default to reading the derived text on the
+  /// [MenuItemButton] itself, which is not guaranteed to be readable.
```
```suggestion
  /// A screen reader will default to reading the derived text on the
  /// [MenuItemButton] itself, which is not guaranteed to be readable.
  /// (For some shortcuts, such as comma, semicolon, and other
  /// punctuation, screen readers read silence)
```

gspencergoog reviewed (approved):
![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png)

philipfranchi committed:
Update packages/flutter/lib/src/material/menu_anchor.dart

Co-authored-by: Greg Spencer <gspencergoog@users.noreply.github.com>

philipfranchi committed:
Merge branch 'master' into add_semantics_label_to_menu_item_button

philipfranchi committed:
Merge branch 'master' into add_semantics_label_to_menu_item_button

gspencergoog commented:
@franciscojma86 Can you give this its second review approval so it can be submitted?

cbracken reviewed (approved):
lgtm - I'd de-bracket the bracketed text after end of sentence in the doc comment, and just make it its own sentence with a period at the end. (Admittedly, I write text like that all the time though)
